src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,523 @@
+		comp_info(dev, "  nr_mode1_pp_param53  %d",
+			  cd->config.igo_params.nr_mode1_pp_param53);
+	}
+}
+
+/* copy and process stream data from source to sink buffers */
+static int32_t igo_nr_copy(struct comp_dev *dev)
+{
+	struct comp_copy_limits cl;
+	struct comp_buffer *sourceb;
+	struct comp_buffer *sinkb;
+	struct igo_nr_comp_data *cd = comp_get_drvdata(dev);
+	int32_t src_frames;
+	int32_t sink_frames;
+
+	comp_dbg(dev, "igo_nr_copy()");
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
+				source_list);
+
+	/* Check for changed configuration */
+	if (comp_is_new_data_blob_available(cd->model_handler))
+		igo_nr_get_igo_params(dev);
+
+	/* Get source, sink, number of frames etc. to process. */
+	comp_get_copy_limits(sourceb, sinkb, &cl);
+
+	src_frames = audio_stream_get_avail_frames(&sourceb->stream);
+	sink_frames = audio_stream_get_free_frames(&sinkb->stream);
+
+	comp_dbg(dev, "src_frames = %d, sink_frames = %d.", src_frames, sink_frames);
+
+	/* Run the data consume/produce below */
+	src_frames = MIN(src_frames, IGO_FRAME_SIZE - cd->in_wpt);
+	sink_frames = MIN(sink_frames, IGO_FRAME_SIZE - cd->out_rpt);
+
+	comp_dbg(dev, "consumed src_frames = %d, produced sink_frames = %d.",
+		 src_frames, sink_frames);
+
+	igo_nr_process(dev, sourceb, sinkb, src_frames, sink_frames,
+		       src_frames * cl.source_frame_bytes,
+		       sink_frames * cl.sink_frame_bytes);
+
+    /* Run algorithm if buffers are available */
+	if ((cd->in_wpt == IGO_FRAME_SIZE) && (cd->out_rpt == IGO_FRAME_SIZE) {
+		comp_dbg(dev, "algo run");
+
+		if (cd->config.igo_params.nr_bypass) {
+			memcpy_s(cd->out, IGO_FRAME_SIZE * sizeof(int16_t),
+				 cd->in, IGO_FRAME_SIZE * sizeof(int16_t));
+		} else {
+			cd->igo_stream_data_in.data = cd->in;
+			cd->igo_stream_data_in.data_width = IGO_DATA_16BIT;
+			cd->igo_stream_data_in.sample_num = IGO_FRAME_SIZE;
+			cd->igo_stream_data_in.sampling_rate = 16000;
+
+			cd->igo_stream_data_ref.data = NULL;
+			cd->igo_stream_data_ref.data_width = 0;
+			cd->igo_stream_data_ref.sample_num = 0;
+			cd->igo_stream_data_ref.sampling_rate = 0;
+
+			cd->igo_stream_data_out.data = cd->out;
+			cd->igo_stream_data_out.data_width = IGO_DATA_16BIT;
+			cd->igo_stream_data_out.sample_num = IGO_FRAME_SIZE;
+			cd->igo_stream_data_out.sampling_rate = 16000;
+
+			IgoLibProcess(cd->p_handle,
+				      &cd->igo_stream_data_in,
+				      &cd->igo_stream_data_ref,
+				      &cd->igo_stream_data_out);
+		}
+
+		cd->in_wpt = 0;
+		cd->out_rpt = 0;
+	}
+
+	return 0;
+}
+
+static int32_t igo_nr_prepare(struct comp_dev *dev)
+{
+	struct igo_nr_comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sourceb;
+	struct comp_buffer *sinkb;
+	int32_t ret;
+
+	comp_info(dev, "igo_nr_prepare()");
+
+	igo_nr_get_igo_params(dev);
+
+	cd->igo_lib_config.algo_name = "igo_nr";
+	cd->igo_lib_config.in_ch_num = 1;
```

Comment:
please don't implicitly assume you are getting one channel, please check the current params against what you support and reject anything not valid so we dont run into corrupted data

Reply:
These are used to configure library as well. Currently the library only support single channel because mono processing takes 250MCPS already.

Reply:
yes, that is fine, but SOF can try and open the component with an invalid format/channels/rate, even if your topology specifies certain config, do not assume that the next configuration engineer will make the same wise decisions. The code should defend against invalid configs.

Reply:
Replied above. Now the wrapper would reject anything other than sample rate = 16000 and s16/s32 in `set_capture_func`.